### PR TITLE
Feature/veg 180 report disposed items

### PR DESCRIPTION
**Pull request by lkeim**
Report works analogously to the "What was purchased" report (see corresponding requirement), except that it does not list sold items, but items that were taken out as spoiled (see corresponding requirement)

### CODE REVIEW

File: src/main/java/de/hsrm/vegetables/my_food_coop_service/controller/ReportsController.java [hunk 1]
```diff
@@ -41,6 +44,15 @@
...
 
+    @NonNull
+    private DisposedRepository disposedRepository;
+
+    @NonNull
+    private DisposeService disposeService;
+
+
+
+
```
Comment (Oxedos) [inline src/main/java/de/hsrm/vegetables/my_food_coop_service/controller/ReportsController.java:55]: A few too much empty lines here, one is enough

File: src/main/java/de/hsrm/vegetables/my_food_coop_service/controller/ReportsController.java [hunk 2]
```diff
@@ -41,6 +44,15 @@
     @NonNull
     private UserService userService;
 
+    @NonNull
+    private DisposedRepository disposedRepository;
```
Comment (Oxedos) [inline src/main/java/de/hsrm/vegetables/my_food_coop_service/controller/ReportsController.java:48]: Do not use repositories in controllers. A repository is solely controller by a service - the gateway to the database.

File: src/main/java/de/hsrm/vegetables/my_food_coop_service/controller/ReportsController.java [hunk 3]
```diff
@@ -153,4 +165,72 @@
...
+
+            float totalVat = StockService.round(disposedDto.getPricePerUnit() * disposedDto.getVat() * disposedDto.getAmount(), 2);
+            disposedItem.setTotalVat(totalVat);
+
+            disposedItem.setGrossAmount(StockService.round(totalVat + disposedDto.getPricePerUnit() * disposedDto.getAmount(), 2));
+            disposedItems.add(disposedItem);
+        });
+
+        return disposedItems;
+    }
```
Comment (Oxedos) [inline src/main/java/de/hsrm/vegetables/my_food_coop_service/controller/ReportsController.java:235]: That's a service function, and it should be moved to the StockService. That has the nice benefit that you don't need to autowire the DisposedRepository into this controller.

Additionally, the sanity check for the timestamps (lines 173 - 179) is business logic and should be done in the service as well (e.g. in this function)

File: src/main/java/de/hsrm/vegetables/my_food_coop_service/controller/ReportsController.java [hunk 4]
```diff
@@ -153,4 +165,72 @@
...
+            disposedItem.setStockId(disposedDto.getStockDto().getId());
+            disposedItem.setPricePerUnit(disposedDto.getPricePerUnit());
+            disposedItem.setUnitType(disposedDto.getUnitType());
+            disposedItem.setVat(StockService.round(disposedDto.getVat(), 2));
+
+            float totalVat = StockService.round(disposedDto.getPricePerUnit() * disposedDto.getVat() * disposedDto.getAmount(), 2);
+            disposedItem.setTotalVat(totalVat);
+
+            disposedItem.setGrossAmount(StockService.round(totalVat + disposedDto.getPricePerUnit() * disposedDto.getAmount(), 2));
+            disposedItems.add(disposedItem);
```
Comment (Oxedos) [inline src/main/java/de/hsrm/vegetables/my_food_coop_service/controller/ReportsController.java:231]: This function is already implemented here in the [StockMapper::disposedDtoToDisposedItem](https://github.com/HSRM-Vegetables/myfoodcoop-service/blob/8bac7ab97f71dc53f507091ec04d052c1fb575ca/src/main/java/de/hsrm/vegetables/my_food_coop_service/mapper/StockMapper.java#L45) and should be used / extended.

That also allows you to then write a function that maps `List<DisposedDto>` to `List<DisposedItem>` similar to this one: [StockMapper::listStockDtoToListStockResponse](https://github.com/HSRM-Vegetables/myfoodcoop-service/blob/8bac7ab97f71dc53f507091ec04d052c1fb575ca/src/main/java/de/hsrm/vegetables/my_food_coop_service/mapper/StockMapper.java#L38)

File: src/main/java/de/hsrm/vegetables/my_food_coop_service/repositories/DisposedRepository.java [hunk 5]
```diff
@@ -1,10 +1,17 @@
...
 import org.springframework.data.jpa.repository.JpaRepository;
 import org.springframework.stereotype.Repository;
 
+import java.time.OffsetDateTime;
+import java.util.List;
+
 @Repository
 public interface DisposedRepository extends JpaRepository<DisposedDto, Long> {
 
+    List<DisposedDto> findAllByUserDto(UserDto userDto);
```
Comment (Oxedos) [inline src/main/java/de/hsrm/vegetables/my_food_coop_service/repositories/DisposedRepository.java:14]: Is this function used anywhere? If not please remove it

File: src/main/java/de/hsrm/vegetables/my_food_coop_service/services/DisposeService.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+import org.springframework.stereotype.Service;
+
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+@Service
+@RequiredArgsConstructor(onConstructor = @__({@Autowired}))
+public class DisposeService {
```
Comment (Oxedos) [inline src/main/java/de/hsrm/vegetables/my_food_coop_service/services/DisposeService.java:18]: I'd call this Dispose**d**Service, as the DTOs are called Dispose**d**Dto as well

File: src/main/java/de/hsrm/vegetables/my_food_coop_service/services/DisposeService.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+@Service
+@RequiredArgsConstructor(onConstructor = @__({@Autowired}))
```
Comment (Oxedos) [inline src/main/java/de/hsrm/vegetables/my_food_coop_service/services/DisposeService.java:17]: That's unnecessary, as there are no autowired class fields

File: src/main/java/de/hsrm/vegetables/my_food_coop_service/services/DisposeService.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+public class DisposeService {
+
+    public static List<VatDetailItem> getVatDetails(List<DisposedItem> disposedItems) {
+        // Get all distinct vat rates
+        ArrayList<Float> distinctVatRates = new ArrayList<>();
+
+        disposedItems.forEach(soldItem -> {
```
Comment (Oxedos) [inline src/main/java/de/hsrm/vegetables/my_food_coop_service/services/DisposeService.java:24]: should read `disposedItem -> {`

**Review by Oxedos — CHANGES_REQUESTED** [review]
Functionality is there, but you need to adhere to our architecture (e.g. what's the responsibility of a service in contrast to a controller)

You're also lacking tests, please add some

File: src/main/java/de/hsrm/vegetables/my_food_coop_service/repositories/DisposedRepository.java (hunk 5 above)
Comment (lkeim) [reply src/main/java/de/hsrm/vegetables/my_food_coop_service/repositories/DisposedRepository.java:14]: used in getDisposedItems()

File: src/main/java/de/hsrm/vegetables/my_food_coop_service/services/DisposeService.java (hunk 7 above)
Comment (lkeim) [reply src/main/java/de/hsrm/vegetables/my_food_coop_service/services/DisposeService.java:17]: but in getVatDetails() these collector is used

Comment (lkeim) [reply src/main/java/de/hsrm/vegetables/my_food_coop_service/services/DisposeService.java:17]: @Service
@RequiredArgsConstructor(onConstructor = @__({@Autowired}))
can be deleted i would agree

File: src/main/java/de/hsrm/vegetables/my_food_coop_service/repositories/DisposedRepository.java (hunk 5 above)
Comment (Oxedos) [reply src/main/java/de/hsrm/vegetables/my_food_coop_service/repositories/DisposedRepository.java:14]: Ah, didn't see that reference 👍

File: src/main/java/de/hsrm/vegetables/my_food_coop_service/services/DisposeService.java (hunk 7 above)
Comment (Oxedos) [reply src/main/java/de/hsrm/vegetables/my_food_coop_service/services/DisposeService.java:17]: The @Service needs to stay, but line 17 needs to be removed

File: src/main/java/de/hsrm/vegetables/my_food_coop_service/services/StockService.java [hunk 9]
```diff
@@ -341,6 +344,62 @@ public DisposedDto dispose(String stockId, UserDto userDto, float amount) {
...
+     */
+    public List<DisposedDto> getDisposedDtos(LocalDate fromDate, LocalDate toDate) {
+        LocalDate today = LocalDate.now();
+        if (fromDate.isAfter(today) || toDate.isAfter(today)) {
+            throw new BadRequestError("Report Date cannot be in the future", ErrorCode.REPORT_DATA_IN_FUTURE);
+        }
+
+        if (fromDate.isAfter(toDate)) {
+            throw new BadRequestError("fromDate cannot be after toDate", ErrorCode.TO_DATE_AFTER_FROM_DATE);
+        }
```
Comment (Oxedos) [inline src/main/java/de/hsrm/vegetables/my_food_coop_service/services/StockService.java:362]: Use the Util function

**Review by Oxedos — CHANGES_REQUESTED** [review]

**Review by Oxedos — APPROVED** [review]

**Pull request merged**